parser/src/parser/grammar.lalrpop

```diff
@@ -367,6 +385,29 @@ CommaElems<T>: Vec<T> = {
     }
 }
 
+SelectorExpr: Expression = {
+    <selector1: SelectorExpr> "&" <selector2: Selector> =>
+        Expression::Mul(Box::new(selector1), Box::new(selector2)),
+    <selector1: SelectorExpr> "|" <selector2: Selector> => {
+        Expression::Sub(
+            Box::new(Expression::Add(
+                Box::new(selector1.clone()), Box::new(selector2.clone()))
+            ),
+            Box::new(Expression::Mul(
+                Box::new(selector1), Box::new(selector2))
+            )
+        )
+    },
+    Selector
+}
+
+Selector: Expression = {
+    "(" <SelectorExpr> ")",
+    <selector: Identifier> => Expression::Elem(selector),
+    "!" <selector: Identifier> =>
+        Expression::Sub(Box::new(Expression::Const(1)), Box::new(Expression::Elem(selector))),
```

Comment:
note: we don't really have a way of checking that this element is binary, but with this transformation we lose the knowledge that we're expecting a binary value. I wonder if we should do something else here. This is true for all of the selectors, really. We should probably enforce that we expect them to be binary, but we won't have this information in the IR. 

Maybe the best thing to do would be to create a new type of expression with `&`, `|`, `!` and let the IR handle rather than the parser.

@bobbinth do you have thoughts on this?

Reply:
Or I guess another option in the IR would be to enforce that every element (or trace reference, etc) used inside one of these expressions as a selector is a binary element, that none of these refer to random values, etc etc. Alternatively, we could just not enforce anything.

Reply:
I was thinking we don't enforce this. I'm not sure we can enforce this at the IR level either? Maybe in the future we could throw a warning in the future if we see these constraints are defined without enforcing a constraint that the selectors used are binary?

Reply:
Yeah, it's easier not to enforce this. Let's leave it for now.

Reply:
But maybe we should open an issue for discussion?

Reply:
Added #219.